File: clients/algoliasearch-client-kotlin/client/src/commonMain/kotlin/com/algolia/client/configuration/ClientOptions.kt
package com.algolia.client.configuration

import com.algolia.client.transport.Requester
import io.ktor.client.*
import io.ktor.client.engine.*
import io.ktor.client.plugins.logging.*
import kotlinx.serialization.json.JsonBuilder
import kotlin.time.Duration
import kotlin.time.Duration.Companion.seconds

public expect class ClientOptions(
  connectTimeout: Duration = 2.seconds,
  writeTimeout: Duration = 30.seconds,
  readTimeout: Duration = 5.seconds,
  logLevel: LogLevel = LogLevel.NONE,
  logger: Logger = Logger.DEFAULT,
  hosts: List<Host>? = null,
  defaultHeaders: Map<String, String>? = null,
  engine: HttpClientEngine? = null,
  httpClientConfig: ((HttpClientConfig<*>) -> Unit)? = null,
  jsonConfig: ((JsonBuilder) -> Unit)? = null,
  requester: Requester? = null,
  algoliaAgentSegments: List<AgentSegment> = emptyList()
) {

  /** Connect timeout for each request */
  public val connectTimeout: Duration

  /** The timeout for each request when performing write operations. */
  public val writeTimeout: Duration

  /** The timeout for each request when performing read operations. */
  public val readTimeout: Duration

  /** [LogLevel] to display in the console. */
  public val logLevel: LogLevel

  /** [Logger] to use for logs. */
  public val logger: Logger

  /** Custom list of hosts and back-up host used to perform a custom retry logic. */
  public val hosts: List<Host>?

  /** Optional default headers that should be applied to every request. */
  public val defaultHeaders: Map<String, String>?

  /** An optional [HttpClientEngine] to specify which HttpEngine should be used by Ktor. */
  public val engine: HttpClientEngine?

  /** An optional [HttpClientConfig] used by Ktor for advanced HttpClient configuration. */
  public val httpClientConfig: ((HttpClientConfig<*>) -> Unit)?

  /** An optional [JsonBuilder] configuration. */
  public val jsonConfig: ((JsonBuilder) -> Unit)?

  /** Custom Http Requester. */
  public val requester: Requester?

  /** List of Algolia agent segments */
  public val algoliaAgentSegments: List<AgentSegment>
}

File: clients/algoliasearch-client-dart/packages/client_core/lib/src/config/client_options.dart
import 'dart:core';

import 'package:algolia_client_core/src/config/agent_segment.dart';
import 'package:algolia_client_core/src/config/host.dart';
import 'package:algolia_client_core/src/transport/requester.dart';

final class ClientOptions {
  /// The list of hosts that the client can connect to.
  final Iterable<Host>? hosts;

  /// The maximum duration to wait for a connection to establish before timing out.
  final Duration connectTimeout;

  /// The maximum duration to wait for a write operation to complete before timing out.
  final Duration writeTimeout;

  /// The maximum duration to wait for a read operation to complete before timing out.
  final Duration readTimeout;

  /// Default headers to include in each HTTP request.
  final Map<String, dynamic>? headers;

  /// List of agent segments for the Algolia service.
  final Iterable<AgentSegment>? agentSegments;

  /// Custom logger for http operations.
  final Function(Object?)? logger;

  /// Custom requester used to send HTTP requests.
  final Requester? requester;

  /// Constructs a [ClientOptions] instance with the provided parameters.
  const ClientOptions({
    this.connectTimeout = const Duration(seconds: 2),
    this.writeTimeout = const Duration(seconds: 30),
    this.readTimeout = const Duration(seconds: 5),
    this.hosts,
    this.headers,
    this.agentSegments,
    this.requester,
    this.logger,
  });

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is ClientOptions &&
          runtimeType == other.runtimeType &&
          hosts == other.hosts &&
          connectTimeout == other.connectTimeout &&
          writeTimeout == other.writeTimeout &&
          readTimeout == other.readTimeout &&
          headers == other.headers &&
          agentSegments == other.agentSegments &&
          requester == other.requester;

  @override
  int get hashCode =>
      hosts.hashCode ^
      connectTimeout.hashCode ^
      writeTimeout.hashCode ^
      readTimeout.hashCode ^
      headers.hashCode ^
      agentSegments.hashCode ^
      requester.hashCode;

  @override
  String toString() {
    return 'ClientOptions{hosts: $hosts, connectTimeout: $connectTimeout, writeTimeout: $writeTimeout, readTimeout: $readTimeout, headers: $headers, agentSegments: $agentSegments, requester: $requester}';
  }
}
